File: selenium_web_tests/pages/login_page.js
const BasePage = require('./base_page');

class LoginPage extends BasePage {
  constructor(driver) {
    super(driver);
    this.usernameInput = { name: 'username' };
    this.passwordInput = { name: 'password' };
    this.submitBtn = { css: 'button[type="submit"]' };
  }

  async login(username, password) {
    await this.navigateTo('/auth/login');
    await this.type(this.usernameInput, username);
    await this.type(this.passwordInput, password);
    await this.click(this.submitBtn);
    return true;
  }
}

module.exports = LoginPage;

File: appium_tests/pages/login_page.py
from pages.base_page import BasePage, By

class LoginPage(BasePage):
    """
    Page Object for Authentication / Login Screen.
    """
    USERNAME_INPUT = (By.NAME, "username")
    PASSWORD_INPUT = (By.NAME, "password")
    LOGIN_BUTTON = (By.XPATH, "//button[@type='submit']")
    ERROR_MESSAGE = (By.CLASS_NAME, "error-msg")

    def login(self, username, password):
        if self.is_mock():
            response = self.driver.post('/auth/login', data={'username': username, 'password': password})
            return response
        else:
            self.driver.get(f"{self.driver.base_url}/auth/login")
            self.send_keys(*self.USERNAME_INPUT, username)
            self.send_keys(*self.PASSWORD_INPUT, password)
            self.click(*self.LOGIN_BUTTON)
            return True

    def get_error_message(self):
        if self.is_mock():
            if "Invalid credentials" in getattr(self.driver, 'page_source', ''):
                return "Invalid credentials"
            return ""
        return self.get_text(*self.ERROR_MESSAGE)
